File: app/src/main/java/sfllhkhan95/doodle/core/models/tools/ToolFactory.java
package sfllhkhan95.doodle.core.models.tools;

import java.lang.reflect.Constructor;

import sfllhkhan95.doodle.core.models.PaintBrush;

/**
 * Created by saifkhichi96 on 23/10/2017.
 */
public class ToolFactory {

    public static Tool get(Class<? extends Tool> type, PaintBrush paintBrush) {
        Tool tool = null;
        try {
            Constructor<? extends Tool> ctor = type.getDeclaredConstructor(PaintBrush.class);
            tool = ctor.newInstance(paintBrush);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return tool;
    }

}

File: app/src/main/java/sfllhkhan95/doodle/core/models/tools/ToolFactory.kt
package sfllhkhan95.doodle.core.models.tools

import sfllhkhan95.doodle.core.models.PaintBrush

/**
 * Created by saifkhichi96 on 23/10/2017.
 */
object ToolFactory {

    operator fun get(type: Class<out Tool>, paintBrush: PaintBrush): Tool? {
        var tool: Tool? = null
        try {
            val ctor = type.getDeclaredConstructor(PaintBrush::class.java)
            tool = ctor.newInstance(paintBrush)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return tool
    }

}
